src/neo/Ledger/MemoryPool.cs

```diff
@@ -253,48 +265,63 @@ internal bool CanTransactionFitInPool(Transaction tx)
         }
 
         /// <summary>
-        /// Adds an already verified transaction to the memory pool.
-        ///
-        /// Note: This must only be called from a single thread (the Blockchain actor). To add a transaction to the pool
-        ///       tell the Blockchain actor about the transaction.
+        /// Verifies (against currentSnapshot) and adds transaction to the memory pool.
         /// </summary>
         /// <param name="hash"></param>
         /// <param name="tx"></param>
         /// <returns></returns>
-        internal VerifyResult TryAdd(Transaction tx, StoreView snapshot)
+        internal VerifyResult TryAdd(Transaction tx)
         {
             var poolItem = new PoolItem(tx);
-
-            if (_unsortedTransactions.ContainsKey(tx.Hash)) return VerifyResult.AlreadyExists;
-
             List<Transaction> removedTransactions = null;
-            _txRwLock.EnterWriteLock();
+            VerifyResult result = VerifyResult.Succeed;
+
+            _snapshotLock.EnterReadLock();
             try
             {
-                VerifyResult result = tx.VerifyStateDependent(snapshot, VerificationContext);
-                if (result != VerifyResult.Succeed) return result;
+                if (currentSnapshot.ContainsTransaction(tx.Hash))
+                    return VerifyResult.AlreadyExists;
 
-                _unsortedTransactions.Add(tx.Hash, poolItem);
-                VerificationContext.AddTransaction(tx);
-                _sortedTransactions.Add(poolItem);
+                result = tx.Verify(currentSnapshot, null);
+                if (result != VerifyResult.Succeed)
+                    return result;
 
-                if (Count > Capacity)
-                    removedTransactions = RemoveOverCapacity();
+                _txRwLock.EnterWriteLock();
```

Comment:
Another problem is that all concurrent threads get stuck here and become single threads. Concurrent threads will lose their effect.

Reply:
It's inevitable, you can only minimize this critical section, but it'll still be there.